src/lib/locals.js

```diff
@@ -13,10 +13,10 @@ const {
 const { generateNonce } = require("./strings");
 
 module.exports = {
-  setLocals: function (req, res, next) {
+  setLocals: async function (req, res, next) {
     res.locals.uaContainerId = GTM_ID;
     res.locals.ga4ContainerId = GTM_ID_GA4;
-    res.locals.cspNonce = generateNonce();
+    res.locals.cspNonce = await generateNonce();
```

Comment:
Time to expose my poor async skills...

If we're awaiting this method, won't execution block until it returns?

Reply:
OK, so reading up, await within an async method will block, but the execution from where the method is called will continue.

In that case, how can we be sure that the nonce will be generated by the time the response is generated?

Reply:
It's actually a bit cleverer than that - `async`/`await` is (approximately) just syntactic sugar for usual promise-based execution.

When you hit an `await` it suspends execution of that event - this is not the same as blocking, because it allows the event loop to pick up other business (i.e. other requests). Once the promise completes, it will queue another task to resume execution of the event from the point that it was awaiting, which will be picked up when the event loop is free.

It might be a bit more obvious what's happening if we rewrote this to use promises:
```
setLocals: function (req, res, next) {
    res.locals.uaContainerId = GTM_ID;
    res.locals.ga4ContainerId = GTM_ID_GA4;
    (generateNonce()).then((nonce) => {
        res.locals.cspNonce = nonce;
        res.locals.isGa4Disabled = GA4_DISABLED;
        res.locals.isUaDisabled = UA_DISABLED;
        res.locals.dynatraceRumUrl = DT_RUM_URL;
        ...
        next();
    });
```
This makes it clear that everything after the `await` is actually run in a separate context (including `next()` - which is what tells express to continue processing the request)

Reply:
OK - so the flow of execution for that particular request wouldn't continue on to the next middleware until the nonce has been generated? The other business that the event loop would pick up would be whole new requests coming in?

Reply:
Contrast this to async code in Java:
```
CompleteableFuture<Object> future = dispatchSomeAsyncTask();

Object value = future.get();
```
The `get()` call here _is_ blocking - the current thread will block until the result of the future is available. The upside is that in Java you can have multiple threads, so blocking isn't quite as fatal as it is in JS

Reply:
> OK - so the flow of execution for that particular request wouldn't continue on to the next middleware until the nonce has been generated? The other business that the event loop would pick up would be whole new requests coming in?
Yeah exactly

Reply:
I'm assuming operations like this are reasonably quick to actually complete given the need te happen during the request/response cycle - I'm now wondering why making them async has made such a large perf difference. Maybe here isn't the place to have that discussion though!

Reply:
I'll message you